lib/modules/manager/bazel-module/rules.ts

```diff
@@ -0,0 +1,149 @@
+import type { SkipReason } from '../../../types';
+import { regEx } from '../../../util/regex';
+import { BazelDatasource } from '../../datasource/bazel';
+import { GithubTagsDatasource } from '../../datasource/github-tags';
+import type { PackageDependency } from '../types';
+import { RecordFragmentSchema, StringFragmentSchema } from './fragments';
+
+// Rule Schemas
+
+export interface BasePackageDep extends PackageDependency {
+  depType: string;
+  depName: string;
+}
+
+export interface OverridePackageDep extends BasePackageDep {
+  // This value is set as the skipReason on the bazel_dep PackageDependency.
+  bazelDepSkipReason: SkipReason;
+}
+
+type BazelModulePackageDep = BasePackageDep | OverridePackageDep;
+
+function isOverride(value: BazelModulePackageDep): value is OverridePackageDep {
+  return (value as OverridePackageDep).bazelDepSkipReason !== undefined;
+}
+
+// This function exists to remove properties that are specific to
+// OverridePackageDep. In theory, there is no harm in leaving the properties
+// as it does not invalidate the PackageDependency interface. However, it might
+// be surprising to someone outside the bazel-module code to see the extra
+// properties.
+export function overrideToPackageDependency(
+  override: Partial<OverridePackageDep>
+): PackageDependency {
+  const copy: Partial<OverridePackageDep> = structuredClone(override);
+  delete copy.bazelDepSkipReason;
+  return copy;
+}
+
+const BazelDepSchema = RecordFragmentSchema.extend({
+  children: z.object({
+    rule: StringFragmentSchema.extend({
+      value: z.literal('bazel_dep'),
+    }),
+    name: StringFragmentSchema,
+    version: StringFragmentSchema,
+  }),
+});
+
+const BazelDepToPackageDep = BazelDepSchema.transform(
+  ({ children: { rule, name, version } }): BasePackageDep => ({
+    datasource: BazelDatasource.id,
+    depType: rule.value,
+    depName: name.value,
+    currentValue: version.value,
+  })
+);
+
+const GitOverrideSchema = RecordFragmentSchema.extend({
+  children: z.object({
+    rule: StringFragmentSchema.extend({
+      value: z.literal('git_override'),
+    }),
+    module_name: StringFragmentSchema,
+    remote: StringFragmentSchema,
+    commit: StringFragmentSchema,
+  }),
+});
+const GitOverrideToPackageDep = GitOverrideSchema.transform(
+  ({
+    children: { rule, module_name: moduleName, remote, commit },
+  }): OverridePackageDep => {
+    const override: OverridePackageDep = {
+      depType: rule.value,
+      depName: moduleName.value,
+      bazelDepSkipReason: 'git-dependency',
+      currentDigest: commit.value,
+    };
+    if (isGithubRemote(remote.value)) {
+      override.datasource = GithubTagsDatasource.id;
+      override.packageName = githubPackageName(remote.value);
+    } else {
+      override.skipReason = 'unsupported-datasource';
+    }
+    return override;
+  }
+);
+
+export const RuleToBazelModulePackageDep = z.union([
+  BazelDepToPackageDep,
+  GitOverrideToPackageDep,
+]);
```

Comment:
Speaking of discriminated unions, I have hypotheses that `type` field will be checked the first, if onlyl it's defined first in the schema. So maybe we don't need to worry about `discriminatedUnion` thing.

Reply:
The fragment structures are tricky. The field that would be useful as a type designator for a discriminated union is in a child structure.

```typescript
{
  children: {
    rule: 'bazel_dep`,  <=== This is the discriminator.
    name: 'rules_foo',
    version: '1.2.3',
  },
}
```

I did not see a way to specify a discriminator that was in a child structure. 

To your point, I am hoping that the `zod` validation logic will quick-exit once it comes across an invalid attribute.